Added `SetCurrentValue`

grokys opened the pull request:
## What does the pull request do?

This PR was developed as a simpler alternative to #9985 and @tomenscape's description there is better than anything I would write so I'll refer you to that PR. The only differences between this PR and #9985 are:

1. Here we don't change the binding priority even when `SetCurrentValue` is called on an `Unset` property value: the priority will still be `Unset`, but in the property diagnostics you will see `IsOverriddenCurrentValue == true` - so no new `Internal` binding priority
2. I've not yet added the change to `TemplateBinding` to use this API in this PR

Not sure whether 2. fits better here or in a separate PR.

I've discussed this with @tomenscape and I believe he agrees that this implementation is simpler, but I would like to wait for his 👍 on this before merging.

grokys committed:
Initial implementation of SetCurrentValue.

grokys committed:
Fix IsSet with SetCurrentValue.

And add unit tests.

grokys committed:
Added untyped SetCurrentValue.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0029916-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

tomenscape commented on src/Avalonia.Base/PropertyStore/ValueStore.cs:
```diff
@@ -159,8 +158,9 @@ public IDisposable AddBinding<T>(DirectPropertyBase<T> property, IObservable<obj
         public void ClearLocalValue(AvaloniaProperty property)
```
This method needs to be renamed now, it sounds like it only has an effect if there is a value with `BindingPriority.Local`.

grokys replied:
Changed to `ClearValue` to match the method on `AvaloniaObject`.

tomenscape commented on src/Avalonia.Base/PropertyStore/ValueStore.cs:
```diff
@@ -209,6 +209,20 @@ public void ClearLocalValue(AvaloniaProperty property)
 
+        public void SetCurrentValue<T>(StyledProperty<T> property, T value)
+        {
+            if (_effectiveValues.TryGetValue(property, out var v))
```
`TryGetEffectiveValue` is used everywhele else, even in `ClearLocalValue`. It seems like a redundant wrapper, though?

grokys replied:
Using `TryGetEffectiveValue` here too now for consistency.

tomenscape commented on src/Avalonia.Base/StyledProperty.cs:
```diff
@@ -220,6 +220,27 @@ internal override void RouteClearValue(AvaloniaObject o)
         }
 
+        [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = TrimmingMessages.ImplicitTypeConvertionSupressWarningMessage)]
+        internal override void RouteSetCurrentValue(AvaloniaObject target, object? value)
```
This method is identical to `RouteSetValue` save for the one line that calls `SetCurrentValue`. The logic should to be extracted to a shared method that returns true if the caller should proceed to set a new value.

grokys replied:
Yeah I noticed you'd refactored it into a separate method; I tend to follow the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming) so didn't do it, but it's not a hard-and-fast rule.

grokys replied:
Refactored this into a separate method. I'm not entirely happy with it as it doesn't meet the rule of three or "methods should be easy to name" rule, but I guess it's not worse than before. I went through a few names for the method but wasn't happy with any of them. Settled on the one that at least reads best.

grokys committed:
ClearLocalValue => ClearValue.

It no longer just clears the local value, and the method on `AvaloniaObject` is called simply `ClearValue` so makes sense to use the same naming here.

grokys committed:
Use TryGetEffectiveValue.

For consistency with the other methods.

grokys committed:
Refactored common code into separate method.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0029989-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into feature/setcurrentvalue

BAndysc commented:
Should this method also solve this problem https://github.com/AvaloniaUI/Avalonia/issues/6684?

maxkatz6 commented:
@BAndysc yes, I think it can be closed now too.

robloo commented:
I switched all the ColorPicker-related controls over to using SetCurrentValue internally. It broke applying the color Palette and the ColorPicker's color value is reset to default when the popup closes. Very strange behavior and it's similar to the original difficult-to-debug binding issues before just duplicating the ColorView template inside ColorPicker.

Regardless, I was not expecting switching to this method to actually break existing template bindings. There might be some secondary effects that have yet to be considered here.

grokys commented:
> There might be some secondary effects that have yet to be considered here.

Yeah quite possibly. Any chance of a minimal repro/unit test that demonstrates the problem?

robloo commented:
>> There might be some secondary effects that have yet to be considered here.
>
> Yeah quite possibly. Any chance of a minimal repro/unit test that demonstrates the problem?

@grokys I opened PR #10439 which right now only changes to using SetCurrentValue(). The issue is visible in the ControlCatalog for the color picker. Note that this is ONLY an issue for the ColorPicker with the Popup. The ColorView and other primitives are working as before.

https://user-images.githubusercontent.com/17993847/220810245-86fb63eb-11c6-4a0b-ac36-e12242572fc9.mp4

Notice:
 1. The ColorPicker control's color is not getting saved properly. It should not be White at first load.
 1. The color in the DropDownButton content is correct as long as the Popup is open. As soon as it is closed it is reset to default.
 1. The Palette bindings never seem to work. Stepping through code these properties are set properly when the control is first created. It is only when the Popup is opened they are destroyed.

I am now thinking this issue might be more related to the Popup lifecycle not behaving as expected with bindings. Perhaps I was actually hiding an issue with the way I'm using bindings by setting the color property directly in the control. Now that the bindings are preserved things are getting reset when the popup opens/closes.

Still not sure how to work-around it though and it has been a problem for this control from the beginning. If someone more knowledgeable tells me nothing is wrong with Popup I'll attempt to find another work-around though.

robloo commented:
Looks like the color picker issues with SetCurrentValue were also fixed by https://github.com/AvaloniaUI/Avalonia/pull/10497.
